packages/protocol/contracts/identity/FederatedAttestations.sol

```diff
@@ -127,38 +181,61 @@ contract FederatedAttestations is
+   * @param maxAttestations Limit the number of attestations that will be returned
+   * @return [0] Array of number of attestations returned per issuer
+   * @return [1 - 3] for m (== sum([0])) found attestations, m <= maxAttestations:
+   *         [
+   *           Array of m accounts,
+   *           Array of m issuedOns,
+   *           Array of m signers
+   *         ]; index corresponds to the same attestation
+   * @dev Adds attestation info to the arrays in order of provided trustedIssuers
+   * @dev Expectation that only one attestation exists per (identifier, issuer, account)
+   */
+  function _lookupAttestations(
     bytes32 identifier,
     address[] memory trustedIssuers,
     uint256 maxAttestations
-  ) public view returns (address[] memory, uint256[] memory, address[] memory) {
-    // Cannot dynamically allocate an in-memory array
-    // For now require a max returned parameter to pre-allocate and then shrink
-    // TODO ASv2 is it a risk to allocate an array of size maxAttestations?
+  ) private view returns (uint256[] memory, address[] memory, uint256[] memory, address[] memory) {
+    uint256[] memory countsPerIssuer = new uint256[](trustedIssuers.length);
+
+    // Pre-computing length of unrevoked attestations requires many storage lookups.
+    // Allow users to call that first and pass this in as maxAttestations.
     // Same index corresponds to same attestation
     address[] memory accounts = new address[](maxAttestations);
     uint256[] memory issuedOns = new uint256[](maxAttestations);
     address[] memory signers = new address[](maxAttestations);
 
     uint256 currIndex = 0;
-
-    for (uint256 i = 0; i < trustedIssuers.length; i = i.add(1)) {
-      uint256 numTrustedIssuers = identifierToAddresses[identifier][trustedIssuers[i]].length;
-      for (uint256 j = 0; j < numTrustedIssuers; j = j.add(1)) {
-        // Only create and push new attestation if we haven't hit max
-        if (currIndex < maxAttestations) {
-          // solhint-disable-next-line max-line-length
-          OwnershipAttestation memory attestation = identifierToAddresses[identifier][trustedIssuers[i]][j];
-          if (!revokedSigners[attestation.signer]) {
-            accounts[currIndex] = attestation.account;
-            issuedOns[currIndex] = attestation.issuedOn;
-            signers[currIndex] = attestation.signer;
-            currIndex = currIndex.add(1);
-          }
-        } else {
-          break;
+    OwnershipAttestation[] memory attestationsPerIssuer;
+
+    for (uint256 i = 0; i < trustedIssuers.length && currIndex < maxAttestations; i = i.add(1)) {
+      attestationsPerIssuer = identifierToAddresses[identifier][trustedIssuers[i]];
+      for (
+        uint256 j = 0;
+        j < attestationsPerIssuer.length && currIndex < maxAttestations;
+        j = j.add(1)
+      ) {
+        if (!revokedSigners[attestationsPerIssuer[j].signer]) {
```

Comment:
Thinking about this now I feel like it might make more sense to just include a boolean `shouldIncludeRevoked` parameter to reduce duplicate code at the smart contract layer. We can always enforce that these methods are called in the correct combinations in the SDK by defining separate functions. What do you think?

Reply:
Yeah, I think this works well with the lookupAll version. Planning to edit to make the following functions available:
- `lookupAttestations` (will probably be renamed) for looking up a bounded number of unrevoked attestations
- `lookupAllAttestations` (will take in a bool to `includeRevoked` or not)
Same thing for identifiers.!

Reply:
Do we have a reason in mind that people may want to lookup revoked attestations? I guess my intuition on revoked attestations is that they are basically dead and should be cleaned up when possible (e.g. the user should have no expectation of being able to reliably access them). Providing this code to access the revoked attestations feels unnecessary to me, and not worth the cost to code size/complexity/gas costs. (One reason I feel this way is that I don't want to hand developers a sharp edge without need. If a developer includes revoked attestations by mistake, it could be a security issue for their app).